In [ ]:
import os
import numpy as np
import torch
import glob
import torch.nn as nn
from torchvision.transforms import transforms
from torch.utils.data import DataLoader
from torch.optim import Adam
from torch.autograd import Variable
import torchvision
import pathlib

In [ ]:
device=torch.device('cuda' if torch.cuda.is_available() else 'cpu')

In [ ]:
print(device)

cuda


In [ ]:
transformer = transforms.Compose([
    transforms.Resize((256, 256)),
    transforms.RandomHorizontalFlip(p=0.5),
    transforms.RandomVerticalFlip(p=0.5),
    transforms.ColorJitter(),
    transforms.ToTensor(),
    transforms.Normalize([0.485, 0.456, 0.406],
                          [0.229, 0.224, 0.225]) 
])

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
train_path="/content/drive/MyDrive/BrainDead_Combined/Train"
test_path="/content/drive/MyDrive/BrainDead_Combined/Test"
 
train_loader = DataLoader(
     torchvision.datasets.ImageFolder(train_path,transform = transformer),
     batch_size = 128, shuffle = True
 )

 
test_loader = DataLoader(
     torchvision.datasets.ImageFolder(test_path,transform = transformer),
     batch_size = 128, shuffle = True
 )

In [ ]:
root = pathlib.Path(train_path)
classes = sorted([j.name.split('/')[-1]for j in root.iterdir()])

In [ ]:
print(classes)

['angry', 'happy', 'neutral', 'sad']


In [ ]:
class ConvNet(nn.Module):
  def __init__(self, num_classes = 5):
    super(ConvNet,self).__init__()

    #First Conv block
    self.conv1 = nn.Conv2d(in_channels=3,out_channels=16, kernel_size=3, stride=1,padding=1)
    self.bn1 = nn.BatchNorm2d(num_features=16)
    self.relu1 = nn.ReLU()
    self.pool1 = nn.MaxPool2d(kernel_size=2)

    #Second Conv Block
    self.conv2 = nn.Conv2d(in_channels=16,out_channels=32, kernel_size=3, stride=1,padding=1)
    self.bn2 = nn.BatchNorm2d(num_features=32)
    self.relu2 = nn.ReLU()
    self.pool2 = nn.MaxPool2d(kernel_size=2)

    self.conv3 = nn.Conv2d(in_channels=32, out_channels=64, kernel_size=3, stride=1,padding=1)
    self.bn3 = nn.BatchNorm2d(num_features=64)
    self.relu3 = nn.ReLU()
    self.pool3 = nn.MaxPool2d(kernel_size=2)

    self.conv4 = nn.Conv2d(in_channels=64, out_channels=128, kernel_size=1, stride=1,padding=1)
    self.bn4 = nn.BatchNorm2d(num_features=128)
    self.relu4 = nn.ReLU()
    self.pool4 = nn.MaxPool2d(kernel_size=2)

    self.fc=nn.Linear(in_features = 128*17*17 ,out_features=256)

    self.fc1=nn.Linear(in_features = 256 ,out_features=num_classes)

  def forward(self,input):
    output=self.conv1(input)
    output=self.bn1(output)
    output=self.relu1(output)
    output=self.pool1(output)

    output=self.conv2(output)
    output=self.bn2(output)
    output=self.relu2(output)
    output=self.pool2(output)
    
    output=self.conv3(output)
    output=self.bn3(output)
    output=self.relu3(output)
    output=self.pool3(output)

    output=self.conv4(output)
    output=self.bn4(output)
    output=self.relu4(output)
    output=self.pool4(output)

    output=output.view(-1,128*17*17)

    output=self.fc(output)
    output=self.fc1(output)
    return output

In [ ]:
model = ConvNet(num_classes=4).to(device)

In [ ]:
optimizer = Adam(model.parameters(), lr=2*0.00001, weight_decay=0.0001)
loss_function = nn.CrossEntropyLoss()

In [ ]:
num_epochs = 100

In [ ]:
train_count = len(glob.glob(train_path+'/**/*.png'))
test_count = len(glob.glob(test_path+'/**/*.png'))

In [ ]:
print(train_count, test_count)

2289 589


In [ ]:
from tqdm import tqdm as tqdm

best_accuracy = 0.0

for epoch in tqdm(range(num_epochs)) :

  model.train()
  train_accuracy=0.0
  train_loss = 0.0

  for i, (images,labels) in enumerate(train_loader):  
    if torch.cuda.is_available():
      images= Variable(images.cuda())
      labels = Variable(labels.cuda())

    optimizer.zero_grad()
    outputs = model(images)
    loss = loss_function(outputs,labels)
    loss.backward()
    optimizer.step()

    train_loss+=loss.cpu().data*images.size(0)
    _,prediction = torch.max(outputs.data,1)

    train_accuracy+=int(torch.sum(prediction==labels.data))

  train_accuracy = train_accuracy/train_count
  train_loss = train_loss/train_count

  model.eval()

  test_accuracy=0.0
  for i, (images,labels) in enumerate(test_loader):  
    if torch.cuda.is_available():
      images= Variable(images.cuda())
      labels = Variable(labels.cuda())

    outputs = model(images)
    _,prediction=torch.max(outputs.data,1)
    test_accuracy+=int(torch.sum(prediction==labels.data))

  test_accuracy = test_accuracy/test_count
  print("\n")
  print('Epoch: '+str(epoch)+' Train Loss: '+str(int(train_loss))+' Train Accuracy: '+str(train_accuracy)+' Test Accuracy: '+str(test_accuracy))
  if test_accuracy>best_accuracy:
    torch.save(model.state_dict(),'best_checkpoint.model')
    best_accuracy = test_accuracy


  1%|          | 1/100 [02:15<3:43:56, 135.72s/it]



Epoch: 0 Train Loss: 1 Train Accuracy: 0.35561380515508956 Test Accuracy: 0.28183361629881154


  2%|▏         | 2/100 [02:43<1:58:05, 72.30s/it] 



Epoch: 1 Train Loss: 1 Train Accuracy: 0.48623853211009177 Test Accuracy: 0.3938879456706282


  3%|▎         | 3/100 [03:11<1:24:19, 52.16s/it]



Epoch: 2 Train Loss: 1 Train Accuracy: 0.5513324595893403 Test Accuracy: 0.4091680814940577


  4%|▍         | 4/100 [03:44<1:11:02, 44.41s/it]



Epoch: 3 Train Loss: 0 Train Accuracy: 0.6020096111839232 Test Accuracy: 0.4397283531409168


  5%|▌         | 5/100 [04:13<1:01:36, 38.91s/it]



Epoch: 4 Train Loss: 0 Train Accuracy: 0.6373962429008301 Test Accuracy: 0.42784380305602715


  6%|▌         | 6/100 [04:42<55:26, 35.39s/it]  



Epoch: 5 Train Loss: 0 Train Accuracy: 0.6404543468763653 Test Accuracy: 0.4855687606112054


  7%|▋         | 7/100 [05:09<50:55, 32.85s/it]



Epoch: 6 Train Loss: 0 Train Accuracy: 0.6640454346876365 Test Accuracy: 0.4838709677419355


  8%|▊         | 8/100 [05:37<47:53, 31.23s/it]



Epoch: 7 Train Loss: 0 Train Accuracy: 0.7155963302752294 Test Accuracy: 0.499151103565365


  9%|▉         | 9/100 [06:04<45:23, 29.93s/it]



Epoch: 8 Train Loss: 0 Train Accuracy: 0.72782874617737 Test Accuracy: 0.5144312393887945


 10%|█         | 10/100 [06:32<43:50, 29.23s/it]



Epoch: 9 Train Loss: 0 Train Accuracy: 0.746177370030581 Test Accuracy: 0.5432937181663837


 11%|█         | 11/100 [06:59<42:35, 28.72s/it]



Epoch: 10 Train Loss: 0 Train Accuracy: 0.7527304499781564 Test Accuracy: 0.5500848896434635


 12%|█▏        | 12/100 [07:27<41:31, 28.31s/it]



Epoch: 11 Train Loss: 0 Train Accuracy: 0.7841852337265182 Test Accuracy: 0.5874363327674024


 13%|█▎        | 13/100 [07:54<40:40, 28.05s/it]



Epoch: 12 Train Loss: 0 Train Accuracy: 0.7776321537789428 Test Accuracy: 0.5636672325976231


 14%|█▍        | 14/100 [08:22<39:59, 27.90s/it]



Epoch: 13 Train Loss: 0 Train Accuracy: 0.7881170816950633 Test Accuracy: 0.5908319185059423


 15%|█▌        | 15/100 [08:49<39:13, 27.69s/it]



Epoch: 14 Train Loss: 0 Train Accuracy: 0.8038444735692443 Test Accuracy: 0.565365025466893


 16%|█▌        | 16/100 [09:17<38:46, 27.70s/it]



Epoch: 15 Train Loss: 0 Train Accuracy: 0.8016601135867191 Test Accuracy: 0.5806451612903226


 17%|█▋        | 17/100 [09:45<38:29, 27.82s/it]



Epoch: 16 Train Loss: 0 Train Accuracy: 0.8235037134119703 Test Accuracy: 0.5483870967741935


 18%|█▊        | 18/100 [10:13<38:12, 27.96s/it]



Epoch: 17 Train Loss: 0 Train Accuracy: 0.8283093053735255 Test Accuracy: 0.6281833616298812


 19%|█▉        | 19/100 [10:41<37:42, 27.93s/it]



Epoch: 18 Train Loss: 0 Train Accuracy: 0.8200087374399301 Test Accuracy: 0.6044142614601019


 20%|██        | 20/100 [11:09<37:14, 27.94s/it]



Epoch: 19 Train Loss: 0 Train Accuracy: 0.8352992573176059 Test Accuracy: 0.6061120543293718


 21%|██        | 21/100 [11:37<36:52, 28.01s/it]



Epoch: 20 Train Loss: 0 Train Accuracy: 0.8409785932721713 Test Accuracy: 0.6044142614601019


 22%|██▏       | 22/100 [12:05<36:16, 27.90s/it]



Epoch: 21 Train Loss: 0 Train Accuracy: 0.854084753167322 Test Accuracy: 0.6213921901528013


 23%|██▎       | 23/100 [12:33<35:50, 27.93s/it]



Epoch: 22 Train Loss: 0 Train Accuracy: 0.8711227610310179 Test Accuracy: 0.6451612903225806


 24%|██▍       | 24/100 [13:01<35:24, 27.95s/it]



Epoch: 23 Train Loss: 0 Train Accuracy: 0.8693752730449978 Test Accuracy: 0.6536502546689303


 25%|██▌       | 25/100 [13:28<34:55, 27.94s/it]



Epoch: 24 Train Loss: 0 Train Accuracy: 0.8768020969855832 Test Accuracy: 0.6129032258064516


 26%|██▌       | 26/100 [13:56<34:20, 27.84s/it]



Epoch: 25 Train Loss: 0 Train Accuracy: 0.8650065530799476 Test Accuracy: 0.6196943972835314


 27%|██▋       | 27/100 [14:24<33:50, 27.81s/it]



Epoch: 26 Train Loss: 0 Train Accuracy: 0.8746177370030581 Test Accuracy: 0.66723259762309


 28%|██▊       | 28/100 [14:52<33:25, 27.86s/it]



Epoch: 27 Train Loss: 0 Train Accuracy: 0.8894713848842289 Test Accuracy: 0.634974533106961


 29%|██▉       | 29/100 [15:19<32:51, 27.77s/it]



Epoch: 28 Train Loss: 0 Train Accuracy: 0.8938401048492791 Test Accuracy: 0.634974533106961


 30%|███       | 30/100 [15:47<32:17, 27.68s/it]



Epoch: 29 Train Loss: 0 Train Accuracy: 0.8934032328527741 Test Accuracy: 0.6095076400679117


 31%|███       | 31/100 [16:14<31:47, 27.64s/it]



Epoch: 30 Train Loss: 0 Train Accuracy: 0.9030144167758847 Test Accuracy: 0.6604414261460102


 32%|███▏      | 32/100 [16:42<31:19, 27.63s/it]



Epoch: 31 Train Loss: 0 Train Accuracy: 0.9034512887723897 Test Accuracy: 0.634974533106961


 33%|███▎      | 33/100 [17:10<30:50, 27.61s/it]



Epoch: 32 Train Loss: 0 Train Accuracy: 0.9025775447793797 Test Accuracy: 0.6621392190152802


 34%|███▍      | 34/100 [17:37<30:20, 27.58s/it]



Epoch: 33 Train Loss: 0 Train Accuracy: 0.9143730886850153 Test Accuracy: 0.6621392190152802


 35%|███▌      | 35/100 [18:05<29:53, 27.59s/it]



Epoch: 34 Train Loss: 0 Train Accuracy: 0.9161205766710354 Test Accuracy: 0.6366723259762309


 36%|███▌      | 36/100 [18:33<29:31, 27.68s/it]



Epoch: 35 Train Loss: 0 Train Accuracy: 0.908256880733945 Test Accuracy: 0.633276740237691


 37%|███▋      | 37/100 [19:00<28:59, 27.61s/it]



Epoch: 36 Train Loss: 0 Train Accuracy: 0.9148099606815203 Test Accuracy: 0.6621392190152802


 38%|███▊      | 38/100 [19:27<28:28, 27.56s/it]



Epoch: 37 Train Loss: 0 Train Accuracy: 0.9314110965487112 Test Accuracy: 0.6468590831918506


 39%|███▉      | 39/100 [19:55<28:01, 27.57s/it]



Epoch: 38 Train Loss: 0 Train Accuracy: 0.9174311926605505 Test Accuracy: 0.634974533106961


 40%|████      | 40/100 [20:22<27:30, 27.50s/it]



Epoch: 39 Train Loss: 0 Train Accuracy: 0.9335954565312363 Test Accuracy: 0.6502546689303905


 41%|████      | 41/100 [20:50<26:59, 27.45s/it]



Epoch: 40 Train Loss: 0 Train Accuracy: 0.9340323285277413 Test Accuracy: 0.6587436332767402


 42%|████▏     | 42/100 [21:17<26:37, 27.53s/it]



Epoch: 41 Train Loss: 0 Train Accuracy: 0.9392747924858017 Test Accuracy: 0.6757215619694398


 43%|████▎     | 43/100 [21:45<26:07, 27.51s/it]



Epoch: 42 Train Loss: 0 Train Accuracy: 0.9379641764962866 Test Accuracy: 0.634974533106961


 44%|████▍     | 44/100 [22:12<25:39, 27.49s/it]



Epoch: 43 Train Loss: 0 Train Accuracy: 0.9322848405417212 Test Accuracy: 0.66723259762309


 45%|████▌     | 45/100 [22:40<25:12, 27.50s/it]



Epoch: 44 Train Loss: 0 Train Accuracy: 0.9161205766710354 Test Accuracy: 0.6621392190152802


 46%|████▌     | 46/100 [23:07<24:43, 27.48s/it]



Epoch: 45 Train Loss: 0 Train Accuracy: 0.9414591524683268 Test Accuracy: 0.66553480475382


 47%|████▋     | 47/100 [23:35<24:17, 27.51s/it]



Epoch: 46 Train Loss: 0 Train Accuracy: 0.944517256443862 Test Accuracy: 0.6451612903225806


 48%|████▊     | 48/100 [24:02<23:49, 27.50s/it]



Epoch: 47 Train Loss: 0 Train Accuracy: 0.9545653123634775 Test Accuracy: 0.6825127334465195


 49%|████▉     | 49/100 [24:30<23:20, 27.46s/it]



Epoch: 48 Train Loss: 0 Train Accuracy: 0.9515072083879423 Test Accuracy: 0.6893039049235993


 50%|█████     | 50/100 [24:57<22:50, 27.40s/it]



Epoch: 49 Train Loss: 0 Train Accuracy: 0.9580602883355177 Test Accuracy: 0.6723259762308998


 51%|█████     | 51/100 [25:24<22:22, 27.40s/it]



Epoch: 50 Train Loss: 0 Train Accuracy: 0.9550021843599825 Test Accuracy: 0.6825127334465195


 52%|█████▏    | 52/100 [25:52<21:52, 27.34s/it]



Epoch: 51 Train Loss: 0 Train Accuracy: 0.9663608562691132 Test Accuracy: 0.6689303904923599


 53%|█████▎    | 53/100 [26:19<21:26, 27.37s/it]



Epoch: 52 Train Loss: 0 Train Accuracy: 0.9611183923110529 Test Accuracy: 0.6876061120543294


 54%|█████▍    | 54/100 [26:46<20:59, 27.38s/it]



Epoch: 53 Train Loss: 0 Train Accuracy: 0.9641764962865881 Test Accuracy: 0.6791171477079796


 55%|█████▌    | 55/100 [27:14<20:31, 27.37s/it]



Epoch: 54 Train Loss: 0 Train Accuracy: 0.9685452162516383 Test Accuracy: 0.6859083191850595


 56%|█████▌    | 56/100 [27:41<20:03, 27.35s/it]



Epoch: 55 Train Loss: 0 Train Accuracy: 0.9541284403669725 Test Accuracy: 0.6825127334465195


 57%|█████▋    | 57/100 [28:08<19:34, 27.32s/it]



Epoch: 56 Train Loss: 0 Train Accuracy: 0.9624290083005679 Test Accuracy: 0.6842105263157895


 58%|█████▊    | 58/100 [28:36<19:09, 27.37s/it]



Epoch: 57 Train Loss: 0 Train Accuracy: 0.9689820882481432 Test Accuracy: 0.6926994906621392


 59%|█████▉    | 59/100 [29:03<18:44, 27.43s/it]



Epoch: 58 Train Loss: 0 Train Accuracy: 0.9659239842726082 Test Accuracy: 0.6825127334465195


 60%|██████    | 60/100 [29:31<18:15, 27.38s/it]



Epoch: 59 Train Loss: 0 Train Accuracy: 0.963739624290083 Test Accuracy: 0.6876061120543294


 61%|██████    | 61/100 [29:58<17:46, 27.35s/it]



Epoch: 60 Train Loss: 0 Train Accuracy: 0.9742245522062036 Test Accuracy: 0.6842105263157895


 62%|██████▏   | 62/100 [30:25<17:18, 27.33s/it]



Epoch: 61 Train Loss: 0 Train Accuracy: 0.9790301441677588 Test Accuracy: 0.6774193548387096


 63%|██████▎   | 63/100 [30:53<16:51, 27.34s/it]



Epoch: 62 Train Loss: 0 Train Accuracy: 0.9733508082131935 Test Accuracy: 0.6960950764006791


 64%|██████▍   | 64/100 [31:20<16:24, 27.34s/it]



Epoch: 63 Train Loss: 0 Train Accuracy: 0.9742245522062036 Test Accuracy: 0.6926994906621392


 65%|██████▌   | 65/100 [31:47<15:57, 27.36s/it]



Epoch: 64 Train Loss: 0 Train Accuracy: 0.9794670161642639 Test Accuracy: 0.6977928692699491


 66%|██████▌   | 66/100 [32:15<15:31, 27.39s/it]



Epoch: 65 Train Loss: 0 Train Accuracy: 0.9777195281782438 Test Accuracy: 0.7062818336162988


 67%|██████▋   | 67/100 [32:42<15:02, 27.35s/it]



Epoch: 66 Train Loss: 0 Train Accuracy: 0.9733508082131935 Test Accuracy: 0.6808149405772496


 68%|██████▊   | 68/100 [33:10<14:36, 27.40s/it]



Epoch: 67 Train Loss: 0 Train Accuracy: 0.9737876802096985 Test Accuracy: 0.6876061120543294


 69%|██████▉   | 69/100 [33:37<14:07, 27.35s/it]



Epoch: 68 Train Loss: 0 Train Accuracy: 0.9764089121887287 Test Accuracy: 0.6825127334465195


 70%|███████   | 70/100 [34:05<13:43, 27.46s/it]



Epoch: 69 Train Loss: 0 Train Accuracy: 0.9746614242027086 Test Accuracy: 0.6910016977928692


 71%|███████   | 71/100 [34:32<13:15, 27.43s/it]



Epoch: 70 Train Loss: 0 Train Accuracy: 0.982088248143294 Test Accuracy: 0.7096774193548387


 72%|███████▏  | 72/100 [34:59<12:48, 27.44s/it]



Epoch: 71 Train Loss: 0 Train Accuracy: 0.9825251201397991 Test Accuracy: 0.6943972835314092


 73%|███████▎  | 73/100 [35:27<12:19, 27.40s/it]



Epoch: 72 Train Loss: 0 Train Accuracy: 0.9803407601572739 Test Accuracy: 0.6859083191850595


 74%|███████▍  | 74/100 [35:54<11:52, 27.40s/it]



Epoch: 73 Train Loss: 0 Train Accuracy: 0.9825251201397991 Test Accuracy: 0.6876061120543294


 75%|███████▌  | 75/100 [36:22<11:26, 27.45s/it]



Epoch: 74 Train Loss: 0 Train Accuracy: 0.9833988641328091 Test Accuracy: 0.7028862478777589


 76%|███████▌  | 76/100 [36:49<10:57, 27.39s/it]



Epoch: 75 Train Loss: 0 Train Accuracy: 0.9855832241153342 Test Accuracy: 0.6808149405772496


 77%|███████▋  | 77/100 [37:16<10:29, 27.39s/it]



Epoch: 76 Train Loss: 0 Train Accuracy: 0.9829619921363041 Test Accuracy: 0.6825127334465195


 78%|███████▊  | 78/100 [37:44<10:01, 27.36s/it]



Epoch: 77 Train Loss: 0 Train Accuracy: 0.9838357361293141 Test Accuracy: 0.6926994906621392


 79%|███████▉  | 79/100 [38:11<09:34, 27.37s/it]



Epoch: 78 Train Loss: 0 Train Accuracy: 0.9825251201397991 Test Accuracy: 0.6926994906621392


 80%|████████  | 80/100 [38:38<09:07, 27.36s/it]



Epoch: 79 Train Loss: 0 Train Accuracy: 0.9882044560943644 Test Accuracy: 0.7096774193548387


 81%|████████  | 81/100 [39:06<08:41, 27.45s/it]



Epoch: 80 Train Loss: 0 Train Accuracy: 0.9851463521188292 Test Accuracy: 0.6910016977928692


 82%|████████▏ | 82/100 [39:33<08:13, 27.41s/it]



Epoch: 81 Train Loss: 0 Train Accuracy: 0.9829619921363041 Test Accuracy: 0.6977928692699491


 83%|████████▎ | 83/100 [40:01<07:45, 27.39s/it]



Epoch: 82 Train Loss: 0 Train Accuracy: 0.9847094801223242 Test Accuracy: 0.6943972835314092


 84%|████████▍ | 84/100 [40:28<07:18, 27.40s/it]



Epoch: 83 Train Loss: 0 Train Accuracy: 0.9886413280908694 Test Accuracy: 0.6994906621392191


 85%|████████▌ | 85/100 [40:55<06:50, 27.37s/it]



Epoch: 84 Train Loss: 0 Train Accuracy: 0.9903888160768894 Test Accuracy: 0.6994906621392191


 86%|████████▌ | 86/100 [41:23<06:23, 27.41s/it]



Epoch: 85 Train Loss: 0 Train Accuracy: 0.9903888160768894 Test Accuracy: 0.6910016977928692


 87%|████████▋ | 87/100 [41:50<05:56, 27.41s/it]



Epoch: 86 Train Loss: 0 Train Accuracy: 0.9921363040629095 Test Accuracy: 0.6893039049235993


 88%|████████▊ | 88/100 [42:17<05:28, 27.36s/it]



Epoch: 87 Train Loss: 0 Train Accuracy: 0.9882044560943644 Test Accuracy: 0.7045840407470289


 89%|████████▉ | 89/100 [42:45<05:01, 27.44s/it]



Epoch: 88 Train Loss: 0 Train Accuracy: 0.9825251201397991 Test Accuracy: 0.7113752122241087


 90%|█████████ | 90/100 [43:15<04:40, 28.06s/it]



Epoch: 89 Train Loss: 0 Train Accuracy: 0.9829619921363041 Test Accuracy: 0.6876061120543294


 91%|█████████ | 91/100 [43:42<04:11, 27.90s/it]



Epoch: 90 Train Loss: 0 Train Accuracy: 0.9899519440803844 Test Accuracy: 0.6910016977928692


 92%|█████████▏| 92/100 [44:10<03:43, 27.89s/it]



Epoch: 91 Train Loss: 0 Train Accuracy: 0.9895150720838795 Test Accuracy: 0.7147707979626485


 93%|█████████▎| 93/100 [44:38<03:15, 27.87s/it]



Epoch: 92 Train Loss: 0 Train Accuracy: 0.9895150720838795 Test Accuracy: 0.7096774193548387


 94%|█████████▍| 94/100 [45:05<02:46, 27.75s/it]



Epoch: 93 Train Loss: 0 Train Accuracy: 0.9903888160768894 Test Accuracy: 0.7045840407470289


 95%|█████████▌| 95/100 [45:33<02:18, 27.72s/it]



Epoch: 94 Train Loss: 0 Train Accuracy: 0.9868938401048493 Test Accuracy: 0.7079796264855688


 96%|█████████▌| 96/100 [46:00<01:50, 27.66s/it]



Epoch: 95 Train Loss: 0 Train Accuracy: 0.9908256880733946 Test Accuracy: 0.7181663837011885


 97%|█████████▋| 97/100 [46:28<01:22, 27.62s/it]



Epoch: 96 Train Loss: 0 Train Accuracy: 0.9930100480559196 Test Accuracy: 0.7079796264855688


 98%|█████████▊| 98/100 [46:55<00:55, 27.57s/it]



Epoch: 97 Train Loss: 0 Train Accuracy: 0.9899519440803844 Test Accuracy: 0.7096774193548387


 99%|█████████▉| 99/100 [47:23<00:27, 27.55s/it]



Epoch: 98 Train Loss: 0 Train Accuracy: 0.9908256880733946 Test Accuracy: 0.7164685908319185


100%|██████████| 100/100 [47:50<00:00, 28.71s/it]



Epoch: 99 Train Loss: 0 Train Accuracy: 0.9934469200524246 Test Accuracy: 0.7079796264855688


Epoch: 99 Train Loss: 0 Train Accuracy: 0.9823204419889503 Test Accuracy: 0.
5745257452574526 on Original Data


Epoch: 99 Train Loss: 0 Train Accuracy: 1.0 Test Accuracy: 0.590909090909091 on Synthetic only


Epoch: 95 Train Loss: 0 Train Accuracy: 0.9908256880733946 Test Accuracy: 0.7181663837011885
 97%|█████████▋| 97/100 [46:28<01:22, 27.62s/it]


In [ ]:
model.to('cuda')

ConvNet(
  (conv1): Conv2d(3, 16, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
  (bn1): BatchNorm2d(16, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  (relu1): ReLU()
  (pool1): MaxPool2d(kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=False)
  (conv2): Conv2d(16, 32, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
  (bn2): BatchNorm2d(32, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  (relu2): ReLU()
  (pool2): MaxPool2d(kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=False)
  (conv3): Conv2d(32, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
  (bn3): BatchNorm2d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  (relu3): ReLU()
  (pool3): MaxPool2d(kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=False)
  (conv4): Conv2d(64, 128, kernel_size=(1, 1), stride=(1, 1), padding=(1, 1))
  (bn4): BatchNorm2d(128, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  (relu4): ReL